### ISSUE TITLE
Login Programmatically?

### ISSUE THREAD

**Issue by rahilwazir** [NodeBB/NodeBB#2428]
How to login in NodeBB programmatically?

I have my main site which is using login system already. I want to automatically login user to NodeBB forum when they log in from the main site. I tried to send post request using `request` module

_I can retrieve CSRF token successfully from API_

```
request.get("http://website.com/forum/api/login", function(err, res, body) {
    var loginUrl, token;
    body = JSON.parse(res.body);
    token = body.token;
    loginUrl = "http://website.com/forum/login?username=" + username + "&password=" + password + "&_csrf=" + token;

    request.post({
      url: loginUrl
    }, function(err, res, body) {
      return console.log(err, res.body);
    });
});
```

Above always return invalid CSRF Token and redirect to 500 error. How to validate CSRF correctly?

**Comment by divinzer** [comment]
is it fixed?
I use 0.7.0 version this problem still happened to me

**Comment by rahilwazir** [comment]
@divinzer Are you in the same case as mine?

**Comment by jarey** [comment]
@ rahilwazir could you post a gist with your working solucion? It might be ver y usefull to other people as a example.
Thanks!

**Comment by agarcia17** [comment]
@rahilwazir it would be awesome if you can the gist of your working solution :)

**Comment by rahilwazir** [comment]
There is no need for gist at all. There is only two things you must have to do.

First you need `csrf_token` from nodebb `/api/config` on document ready.
In my case it was angular, so I ran below code within `.run` block

```
var forum_url = 'http://YOUR_NODEBB_FORM.com/api/config';
var Forum = {};
$http.get(forum_url)
    .success (function(res) { Forum.csrf_token = res.csrf_token; });
```

When user try to login I set the request header `x-csrf-token` to `Forum.csrf_token`
So again in angular it was something like this

```
$http.post ('/api/session', data, {
  headers: { 'x-csrf-token': Forum.csrf_token }
})
```

On server-side I try to login in nodebb when my primary website login succeeds, so it was something like this (in nodejs):

**Comment by rahilwazir** [comment]
There is only two things you must have to do.

First you need `csrf_token` from nodebb `/api/config` on document ready.
In my case it was angular, so I ran below code within `.run` block

```
var forum_url = 'http://YOUR_NODEBB_FORUM.com/api/config';
var Forum = {};
$http.get(forum_url)
    .success (function(res) { Forum.csrf_token = res.csrf_token; });
```

When user try to login on my primary website (through ajax call), I set the request header `x-csrf-token` to `Forum.csrf_token`
So again in angular it was something like this

```
$http.post ('/api/session', loginData, {
  headers: { 'x-csrf-token': Forum.csrf_token // retrieved from the above snippet }
})
```

On server-side I try to login in nodebb when my primary website login succeeds, so it was something like this (in nodejs using `request` module):

```
passport.authenticate("local", function(err, user, info){
  // If my primary site login succeeds here.. then try login call in nodebb

  var nodebb_login = 'http://YOUR_NODEBB_FORUM.com/login';
  var options = {
    form: {
      username: 'nodebb_user_name',
      password: 'nodebb_user_pass',
      remember: 'on'
    },
    headers : {
      'Cookie': req.headers.cookie, // I'm not sure this necessary though
      'X-CSRF-Token': req.headers['x-csrf-token'] // Here I'm setting the same csrf token , which is retrieved from the initial client side call of `/api/config`
    }
  };

  request.post(nodebb_login, options, function(err, res) {
    // forum login result??
  });
});
```

Also remember I was using the same username/password for my nodebb users as they login to my primary website.
I register users to nodebb using `/register` api call, when they register for my primary website. And for `/register` same steps as above have to follow.

**Comment by agarcia17** [comment]
@rahilwazir that is awesome! thank you so much for sharing!

**Comment by thanhse** [comment]
Hello @rahilwazir 
I have got error **invalid csrf token** on version 1.0.3. This is my code. Please advise help me

var forum_url = 'http://forum2.swimoutlet.com:4567/api/config';
            var Forum = {}; 
            var token_all;
            $.ajax({
                url: forum_url,
                dataType: "json",
                type: "GET",
                async: false,
                success: function (err, res, body) {
                    var body = JSON.parse(body.responseText);                    
                    Forum.csrf_token = body.csrf_token;                    
                }

            });

var loginData = {
                    'username': "thanh.huynh",
                    'password': "123456",
                    'remember': true,
                    '_csrf': Forum.csrf_token
                }

$.ajax({
                    url: "http://forum2.swimoutlet.com:4567/login",
                    type: "POST",
                    async: false,
                    headers: { 'Cookie': document.cookie,  'x-csrf-token': Forum.csrf_token },                    
                    data: loginData, 
                    success: function (err, res, body) {
                        alert(err + " " + res + "  " + body.responseText);
                    }
                });     

Thanks.

**Comment by rahilwazir** [comment]
@thanhse Well, it seems like you are trying to login through clientside (Which require CORS), I'm not sure about that. I'm no longer active on this module, so will try to help.

- Make sure you have [CORS](https://github.com/NodeBB/NodeBB/pull/3899) enabled for NodeBB.
- Or you can login through server-side call, which don't require CORS and send request to NodeBB using `curl` (if you are on PHP) or `request` module (if on NodeJS)

**Comment by thanhse** [comment]
Hello rahilwazir 
I tried to login through server-side call. But still got error Forbidden.

 Dim xmlhttp, url
 url = "http://www.spflforum.com/api/config"
set xmlhttp = CreateObject("MSXML2.ServerXMLHTTP") 
xmlhttp.open "GET", url, false    
xmlhttp.send
   
Dim jsonIdentities : 
set jsonIdentities = JSON.parse(xmlhttp.responseText)
    
Dim csrf_token
**csrf_token** = jsonIdentities.get("csrf_token")

Set xmlhttp = nothing
Dim DataToSend
DataToSend = "username=thanh.huynh&password=123456&remember=1&_csrf=" & csrf_token

url = "http://www.spflforum.com/login"
set xmlhttp = CreateObject("MSXML2.ServerXMLHTTP") 
xmlhttp.open "POST", url, false
xmlhttp.setRequestHeader "Content-Type", "application/x-www-form-urlencoded"
xmlhttp.setRequestHeader "Access-Control-Allow-Origin", "*"
xmlhttp.SetRequestHeader "x-csrf-token", csrf_token
xmlhttp.send(DataToSend)

Please advise.
Thanks.

**Comment by rahilwazir** [comment]
I assume it's server-side call. Again it seems like you are doing the wrong approach on server-side call, please see https://github.com/NodeBB/NodeBB/issues/2428#issuecomment-68245771

You need to get the token from client-side call and send it to your login call on server-side. And use that token (no need to retrieve the token within server-side call) for forum login.

**Comment by thanhse** [comment]
Hello @rahilwazir
I have implement step by step as your request's

var forum_url = 'http://www.spflforum.com/api/config';
var Forum = {};
var token_all;
$.ajax({
           url: forum_url,
            dataType: "json",
            type: "GET",
            async: false,
            crossdomain: true,
            success: function (err, res, body) {
                var body = JSON.parse(body.responseText);
                Forum.csrf_token = body.csrf_token;
                $.ajax({
                    url: "http://**www.mysite.com/getlogin.asp**",
                    type: "POST",
                    data: { username: "thanh.huynh", password: "123456" },
                    success: function (err, res, body) {
                    },
                    beforeSend: function (xhr)
                    { xhr.setRequestHeader("x-csrf-token", Forum.csrf_token); }
                });
            }
        });

In **mysite.com/getlogin.asp** page and use token for forum login
Server-side call
 Dim xmlhttp, url, csrf_token, DataToSend
 csrf_token =  Request.ServerVariables("HTTP_x-csrf-token")            
 DataToSend = "username=thanh.huynh&password=123456&remember=1"
 url = "http://www.spflforum.com/login"
 set xmlhttp = CreateObject("MSXML2.ServerXMLHTTP") 
 xmlhttp.open "POST", url, false    
 xmlhttp.setRequestHeader "Content-Type", "application/x-www-form-urlencoded"
 xmlhttp.SetRequestHeader "x-csrf-token", csrf_token
 xmlhttp.send(DataToSend)          

Still got error Forbidden
Thanks.

**Comment by rahilwazir** [comment]
:disappointed: That's sad, @julianlam any input?

**Comment by thanhse** [comment]
Hello @rahilwazir,
Any ideas?

**Comment by rahilwazir** [comment]
@thanhse No, that's the only it should work, I really can't see you are doing anything wrong here

**Comment by thanhse** [comment]
Hello @julianlam 
Please check help me.
Thanks.

**Comment by windht** [comment]
I managed to overcome this **under same domain**.

I did all required steps @rahilwazir mentioned. My own auth site is on `myown.com` and forum is on `forum.anothermyown.com`. This fails. **Forbidden all the times.**

Below are my steps that worked:

1. Use nginx to redirect all "myown.com/forum" (a subfolder like url) to the "forum.anothermywon.com"'s **IP and PORT**. 

2. Change the config.json url `http://127.0.0.1:port/forum`. 

3.Run `./nodebb setup` and then `./nodebb restart` . 

Try it, no more forbidden. Just `window.location.href = "/forum"` in your `myown.com`  login success callback

This is tested in v1.3.0.
